### PR TITLE
Fail shard if IndexShard#storeStats runs into an IOException

### PR DESCRIPTION
**Pull request by andrershov**
Fix for https://github.com/elastic/elasticsearch/issues/29008.
There are two places that I'm not sure about, I've marked them TODO.
Based on previous pull request https://github.com/elastic/elasticsearch/pull/29078, which can be closed if this one goes through.

### CODE REVIEW

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks great. I left some minor comments.

**Comment by colings86** [comment]
@andrershov could you please add labels to this PR (area label, change type label and version label(s) )?

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: @bleskes You mean, that I can just check if shard failure event listener has triggered instead of checking shard state? If I remove shard closing from implementation, then yes, actually shard state would be opened. It was in original proposed patch by you, that's why I've included logic to close the shard.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: @bleskes do you mean that in real application (not in unit tests), IndiciesClusterStateService will realize that shard has failed and will close the shard? That's why close call would no longer be needed.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: @bleskes what do you think about explicitly closing the directory here? I've added this line, because w/o it test fails and reports unclosed resources. But is not it a shard that is responsible for directory lifecycle?

**Comment by andrershov** [comment]
I've fixed formatting and nit things. @bleskes need to get your reply to 3 questions and I'll fix the rest.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java
Comment (bleskes) [reply server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: > @bleskes You mean, that I can just check if shard failure event listener has triggered instead of checking shard state?

yes

>  It was in original proposed patch by you, that's why I've included logic to close the shard.

I told you I wasn't sure of it's quality :)

**Comment by andrershov** [comment]
@bleskes Fixed all the issues: 1) Removed close call for shard 2) Replaced CLOSE state check with checking that failure handler has triggered 3) Replaced explicit directory close call to try with resources for store

**Comment by andrershov** [comment]
@bleskes should it be included in 7.0 only?

**Review by bleskes — APPROVED** [review]
LGTM

**Comment by bleskes** [comment]
@andrershov I think this is fine to into 6.4 as well

File: server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java]: Nit: This is probably better on a separate line like the rest of the parameters.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: Nit: This is probably better on a separate line like the rest of the parameters.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: This should say `storeStats`, not `storeState`.

File: test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java
Comment (jasontedor) [inline test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java]: Nit: This is probably better on a separate line like the rest of the parameters.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a few nits, and caught one error in the shard failure message.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: One more thing, looking at the implementation of `failShard`, we don't need to append `e.getMessage()` here.

**Review by jasontedor — APPROVED** [review]
LGTM. Thanks @andrershov.

**Comment by jasontedor** [comment]
@andrershov Please wait for all CI checks to complete before merging. This is a step we take to reduce the likelihood that we break the build for the rest of the team. 😄

**Pull request merged**
